Kaleido 3 Rainbow Artifact Reduction

utopiafallen opened the pull request:
- fix #771 

The color filter array for Kaleido 3 color displays can sometimes interfere with the dither patterns used in manga for gray tones, causing rainbows to appear in those areas (essentially a form of moire artifacting). While there are more sophisticated solutions, this commit introduces the naive solution of breaking up the dither pattern through blurring the image to fix the rainbow moire.

- Introduced display type information (represented as an enum) to device profiles, determined based on device names.
- Introduced an 'optimizeForDisplay()' processing step to image file processing that happens after resizing (partly for performance but mostly for adjusting the actual displayed image). For devices with Kaleido 3 color displays, this applies unsharp->box blur->unsharp to the image in an attempt to break up the interference pattern without losing too much sharpness in the original image.

axu2 commented:
@oncpnda

axu2 commented:
Out of curiosity how does this compare using a mismatched resolution like the pw11 or scribe profiles?

utopiafallen commented:
Sorry, no idea, didn't try those with my Colorsoft.

neyney10 commented:
Ok. so my main issues:

1. It limits itself to only specific devices with the new method "DisplayTypeFromDeviceName", instead of incorporating the displaytype into the device profiles maps that already exist.
2. I always used "Other" as my device isnt in the list so I set a custom width and height. (regardless i like to play with it sometimes). And it doesnt allow me to enable or disable the Rainbow fix processing action.
3. I'll list what my results:

I selected a manga page that displays the rainbow artifact on my device, an image that I already encountered in the past (using the custom width and height).

Now, i selected the Kobo Libra Color as the device in KCC to test the rainbow artifact reduction, which has lower resolution (width/height) than my custom and device's resolution.

I didnt have the rainbow effect on the "downscaled" image, with, and without the rainbow artifact reduction method. But, the one WITHOUT the fix, had a different issue of visible RGB grid array (of the Kaleido 3 screen),  which is similar to the rainbow effect but different. And the image WITH the fix didnt have this issue, the grid array isnt visible and image looks normal without issues (although a bit small due to the lower resolution compared to my device).

Can you share some images with this rainbow effect on your end so that I can test them on my device?

utopiafallen commented:
![Rainbow1](https://github.com/user-attachments/assets/9ca79a4b-1301-4875-b056-e642c4582c45)
![Rainbow2](https://github.com/user-attachments/assets/1b3eb599-1f92-435d-bd4a-3700b6a3925a)

neyney10 commented:
Sorry. I couldn't replicate the rainbow effect with those images on my device.

axu2 commented:
Sounds like it should be a ui toggle option.

utopiafallen commented:
So...can this pull request be accepted or rejected? Seems like you and
neyney10 know what you would want to do with the code after accepting the
changes. I made the mistake of not making a feature branch for the PR so
I've been holding off on pushing new commits to my master while I wait for
this to resolve one way or another.

Thanks!

On Mon, Feb 3, 2025, 7:38 AM Alex Xu ***@***.***> wrote:

> Sounds like it should be a ui toggle option.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ciromattia/kcc/pull/821#issuecomment-2631357193>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABOT7JOVU5SFCDMX256NEUT2N6EQXAVCNFSM6AAAAABWGYVMIWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMZRGM2TOMJZGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

axu2 commented:
I'll accept it if you make it a ui toggle option. 

The readme has example of adding new checkbox.

utopiafallen commented:
Great, I'll consider it rejected then.

axu2 commented:
Wait, are you planning on opening a new PR in another branch? @utopiafallen 

Is moving the display type into one of the existing maps possible? If so, you could just add a custom (color) profile.

utopiafallen commented:
>Wait, are you planning on opening a new PR in another branch?

No, the actual "fix" is trivial enough that I'm confident you or neyney10 could easily re-implement it according to the project's needs and best practices.

>Is moving the display type into one of the existing maps possible? If so, you could just add an "other (color)" profile.

I personally hate using dynamically typed languages for anything non-trivial so I already kind of hate how all the data is organized and would want a clearer separation between device traits and processing profiles, and rework the processing code and UI to reflect that distinction, as well as data drive both aspects. 

But philosophical grievances aside, the actual reason I'm moving on is my changes meet my current needs so I have no reason to do more, and my next priority is to optimize the processing pipeline so that the only bottlenecks are just extracting source images and processing them. Once I improve that, KCC is feature complete for me for a while 😄

axu2 commented:
OK, I went ahead and added the checkbox, I asked earlier in case you were interested in learning to do it yourself.

The data organization is definitely not the greatest, but it was worse before. I already cleaned up a bit in 

- #764 

Feel free to PR your other optimizations, maybe we'll be able to offer insight.

utopiafallen commented:
Awesome, thank you for getting it over the finish line! I was not
interested in learning 😁

As for the optimizations, feel free to look at my fork for reference. I'm
sure many of the changes are not done to the projects standards.

On Sun, Feb 9, 2025, 7:08 PM Alex Xu ***@***.***> wrote:

> OK, I went ahead and added the checkbox, I asked earlier in case you were
> interested in learning to do it yourself.
>
> The data organization is definitely not the greatest, but it was worse
> before. I already cleaned up a bit in
>
>    - #764 <https://github.com/ciromattia/kcc/pull/764>
>
> Feel free to PR your other optimizations, maybe we'll be able to offer
> insight. @utopiafallen <https://github.com/utopiafallen>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ciromattia/kcc/pull/821#issuecomment-2646820393>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABOT7JICPCWEIW7TUY2TD4T2PAJ2DAVCNFSM6AAAAABWGYVMIWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNBWHAZDAMZZGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

axu2 commented:
@utopiafallen 

There's a lot of interesting commits you've got in your history, and thanks for pointing out that I sanitized twice by accident. I already fixed it in latest release.

The rest of your commits will take a while to get through. 

If you want to, feel free to open a PR with a single `git cherry-pick b0c2aefc4b35f003f2bc64f61c786791a0760c70` commit, that's where I could start. I recently ran into issues if `tar` is ran first and fails.

- https://github.com/utopiafallen/kcc/commit/b0c2aefc4b35f003f2bc64f61c786791a0760c70

I need to ask though, your deleteFilesWithRetries is surprising, did you have that issue before all your changes? I've never had that issue, but maybe it's because the stuff you changed made things too fast so a race occured?

utopiafallen commented:
> I need to ask though, your deleteFilesWithRetries is surprising, did you have that issue before all your changes? I've never had that issue, but maybe it's because the stuff you changed made things too fast so a race occured?
Same, I didn't have this issue until I started working on optimizing the processing, but I don't think it's related to any speedups because my changes would have actually slowed down each individual image processing job and its the process for each job not releasing their file handles despite `join()` finishing that seems to be the problem (and since the retry does usually work, I believe this is just an interaction between Python's multi-process stuff and Windows).

> And for organization, maybe put all your changes on a dev branch and keep master clean?
Yea, I didn't realize this was the expected workflow when it comes to git and Github, but I have no other changes planned at this point so not worth the effort to reorganize.
